### PR TITLE
Add boolean values script fields

### PR DESCRIPTION
**Pull request by nik9000**
This adds `boolean` values `runtime_script` fields.

### CODE REVIEW

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/ScriptBooleanMappedFieldType.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    }
+
+    @Override
+    public Query existsQuery(QueryShardContext context) {
+        checkAllowExpensiveQueries(context);
+        return new BooleanScriptFieldExistsQuery(script, leafFactory(context.lookup()), name());
+    }
+
+    @Override
+    public Query rangeQuery(
```
Comment (nik9000) [inline x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/ScriptBooleanMappedFieldType.java:74]: I'm trying to mimic what we support in core. It works, so far as I can tell. It is kind of amazing we allow this at all, but we do!

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/fielddata/ScriptBooleanDocValues.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    public boolean advanceExact(int docId) {
+        script.runForDoc(docId);
+        cursor = 0;
+        return script.trues() > 0 || script.falses() > 0;
+    }
+
+    @Override
+    public long nextValue() throws IOException {
+        // Emit all false values before all true values
+        return cursor++ < script.falses() ? 0 : 1;
```
Comment (javanna) [inline x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/fielddata/ScriptBooleanDocValues.java:32]: I must say that multiple values for booleans look really weird, especially trying to summarize multiple values into one, and re-sorting them. Is this behaviour documented anywhere?

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/fielddata/ScriptBooleanFieldData.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    @Override
+    public ScriptBooleanLeafFieldData loadDirect(LeafReaderContext context) throws IOException {
+        return new ScriptBooleanLeafFieldData(new ScriptBooleanDocValues(leafFactory.newInstance(context)));
+    }
+
+    @Override
+    public NumericType getNumericType() {
+        return NumericType.DOUBLE;
```
Comment (javanna) [inline x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/fielddata/ScriptBooleanFieldData.java:75]: I never realized this either: booleans are double????

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/ScriptBooleanMappedFieldType.java [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    }
+
+    @Override
+    protected String runtimeType() {
+        return BooleanFieldMapper.CONTENT_TYPE;
+    }
+
+    @Override
+    public Object valueForDisplay(Object value) {
+        throw new UnsupportedOperationException();
```
Comment (javanna) [inline x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/ScriptBooleanMappedFieldType.java:44]: why do we throw? Shouldn't we borrow the logic from https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/mapper/BooleanFieldMapper.java#L157 ?

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/ScriptBooleanMappedFieldType.java (hunk 1 above)
Comment (javanna) [reply x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/ScriptBooleanMappedFieldType.java:74]: oh boy :)

**Review by javanna — APPROVED** [review]
thanks for all the tests, left a couple of comments but LGTM. i was wondering if we should have some duel tests to compare behaviour between concrete boolean fields and runtime boolean fields, given that we discovered some nuances around range queries etc.

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/fielddata/ScriptBooleanDocValues.java (hunk 2 above)
Comment (nik9000) [reply x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/fielddata/ScriptBooleanDocValues.java:32]: I'm not sure it is, though it is what happens when you use doc values for out of the box booleans.

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/fielddata/ScriptBooleanFieldData.java (hunk 3 above)
Comment (nik9000) [reply x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/fielddata/ScriptBooleanFieldData.java:75]: Ooops, copy and paste error. Fixing.

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/ScriptBooleanMappedFieldType.java (hunk 4 above)
Comment (nik9000) [reply x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/ScriptBooleanMappedFieldType.java:44]: I can do that, but I don't think it is ever called!

**Comment by nik9000** [comment]
> i was wondering if we should have some duel tests to compare behaviour between concrete boolean fields and runtime boolean fields, given that we discovered some nuances around range queries etc.

I've added some dual tests. I think it'd probably be good to add them for every field to be honest. I can work on that.

**Comment by nik9000** [comment]
run elasticsearch-ci/2

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/ScriptBooleanMappedFieldType.java (hunk 4 above)
Comment (javanna) [reply x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/ScriptBooleanMappedFieldType.java]: oh well I did not realize it was unused, in that case I am good with throwing, but add a comment?

**Pull request merged**